### PR TITLE
Issue 434: Fixed imprecise `cref`s in XML Docs

### LINKED ISSUE #434: Documentation `cref`s point to nonexisting classes
### System information

- **OS version/distro**: Windows 10 v1803
- **.NET Version (eg., dotnet --info)**: 2.1.201

### Issue

- **What did you do?** I opened the solution in Jetbrains Rider.
- **What happened?** Rider pointed out a great many mistakes in the `cref` elements of XMLDocs
- **What did you expect?** Perfect documentation, of course :-)

### PR DESCRIPTION
**Pull request by markusweimer**
This fixes a couple of dangling `cref` in the XML Docs. This commit doesn't contain functional changes to the code.

Issue:
  This closes #434

### CODE REVIEW

**Comment by markusweimer** [comment]
The test failure seems odd, given that there are no functional changes in this PR.

File: src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs [hunk 1]
```diff
@@ -350,7 +350,7 @@ public static void AddMultWithOffset(ref VBuffer<Float> src, Float c, ref VBuffe
         /// Perform in-place scaling of a vector into another vector as
         /// <c><paramref name="dst"/> = <paramref name="src"/> * <paramref name="c"/></c>.
```
Comment (codemzs) [inline src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs]: I wonder why we have this depricated folder. Can you please check if the functions in here have any references?

Comment (TomFinley) [reply src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs]: They certainly do. I have no idea why this was put into deprecated, maybe @Ivanidzo4ka knows.

Comment (codemzs) [reply src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs]: It is also spelled as "Depricated" :(

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs]: as an responsible adult I would totally push my fault on shoulders of others
[https://msdata.visualstudio.com/AlgorithmsAndDataScience/_git/TLC/commit/f03fff36654042146dbebda077356bf45d534127?refName=refs%2Fheads%2FTLC_Before_github&_a=contents&path=%2FCore%2Fsrc%2FTlcMain%2FTlcMain.csproj](https://msdata.visualstudio.com/AlgorithmsAndDataScience/_git/TLC/commit/f03fff36654042146dbebda077356bf45d534127?refName=refs%2Fheads%2FTLC_Before_github&_a=contents&path=%2FCore%2Fsrc%2FTlcMain%2FTlcMain.csproj)

Let's say it's been like this forever.



---
In reply to: [199894576](https://github.com/dotnet/machinelearning/pull/485#discussion_r199894576) [](ancestors = 199894576)

Comment (TomFinley) [reply src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs]: Sounds fine.

**Comment by markusweimer** [comment]
> I do not object to the change per se, but I wonder if we can have a policy if whether Rider support is actually a goal of this project,

Good point. I'd say no to that. As long as our builds work from the command line on all the platforms we target, we can stay out of the business of recommending or "supporting" tools.

What tool do we use to render API docs? If that tool can properly link the `cref` instances mentioned here, I am ok with closing this as a WONTFIX.

> especially when it appears that the problem is that they have a bug with their doc comment parser. (Or else, VS is too permissive.)

I don't know enough about the rules for XML Docs. From what I understand from the warnings I got, we have two kinds of fixes in this PR:

  * `cref` pointing to a class that wasn't imported: Not raising an issue in VS here seems like overly lenient. How does the reader know which class is being referenced? Are there heuristics to be applied?
  * `cref` missing type parameters: I can see cases here where the type parameters aren't necessary to uniquely identify the class. However, not adding them seems brittle in the face of future change, as ambiguity could be introduced at a later time. And without tooling to raise an alert then, our documentation is at risk of deteriorating.

Hence, I currently believe that this PR is rooted in VS being more permissive than Rider / ReSharper.

**Comment by sharwell** [comment]
@markusweimer helped narrow down the sources of behavior differences. I'm now **neutral** on this pull request (it's implemented correctly even if it's not required by the compiler), but believe that the specific changes in #499 are important to avoid problems as development progresses.

**Comment by shauheen** [comment]
@codemzs are your comments resolved?

**Comment by markusweimer** [comment]
With #499 merged, I'd like to rebase this and address @codemzs's comments at the same time.

**Comment by markusweimer** [comment]
I have done a rebase and addressed @codemzs's comments.

**Comment by markusweimer** [comment]
@codemzs, @TomFinley, @sharwell How shall this proceed? Can it be merged?

**Pull request merged**